### PR TITLE
add bomberman question

### PR DESCRIPTION
**Pull request by jianhaolyu**
add bomber question.
style checked

### CODE REVIEW

File: client/app.html~ [hunk 1]
```diff
@@ -0,0 +1,49 @@
+<!DOCTYPE html>
```
Comment (seanlip) [inline client/app.html~]: Why commit this file?

File: assets/question_sets/strings.js~ [hunk 2]
```diff
@@ -0,0 +1,30 @@
+// Copyright 2017 The TIE Authors. All Rights Reserved.
```
Comment (seanlip) [inline assets/question_sets/strings.js~]: Why include this file? (Make sure to look at the "Files Changed" tab on https://github.com/google/tie/pull/147/files and ensure that the change is what you expect it to be. You can also use tools like "git diff" or "meld".)

File: assets/questions/bomberman.js [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+            for c in range(len(board[0])):
+                if board[r][c] == "e":
+                    count += 1
+        return count
+`
+  },
+  tasks: [{
+    instructions: [
+      [
+        'For this question, you will be given a borad, in which all elements are string. ',
```
Comment (apicardgoog) [inline assets/questions/bomberman.js]: borad -> board

File: assets/questions/bomberman.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        'X represent a wall, e represent an enemy, "" represent an empty cell. ',
```
Comment (apicardgoog) [inline assets/questions/bomberman.js]: represent -> represents

File: assets/questions/bomberman.js [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        'You can place an bomb in an empty cell, which will kill enemies in vertical and horizontal ',
+        'direction until it met a wall. ',
```
Comment (apicardgoog) [inline assets/questions/bomberman.js]: directions until a wall is met.

File: assets/questions/bomberman.js [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+            right = 0
+            for c in range(cols - 1, -1, -1):
+                if board[r][c] == "e":
+                    right += 1
+                    down[c] += 1
+                else:
+                    memo[r][c] += right + down[c]
+                    if memo[r][c] > maxEnemy:
+                        maxEnemy = memo[r][c]
+        return maxEnemy
```
Comment (eyurko) [inline assets/questions/bomberman.js]: Add a newline between methods.

File: assets/questions/bomberman.js [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+`def bomb(board):
+    return 0
+`
+  },
+  auxiliaryCode: {
+    python:
+`class AuxiliaryCode(object):
+    @classmethod
+    def negative(cls, board):
+        return -1
```
Comment (eyurko) [inline assets/questions/bomberman.js]: Add a newline between methods.

File: assets/questions/bomberman.js [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+    @classmethod
+    def ignoreWalls(cls, board):
+        if len(board) == 0 or len(board[0]) == 0: return 0
```
Comment (eyurko) [inline assets/questions/bomberman.js]: if !board or !board[0]. Don't use explicit length comparisons in Python.

File: assets/questions/bomberman.js [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        rows = len(board)
+        cols = len(board[0])
+        maxEnemy = 0
+        up = [0 for x in range(cols)]
+        left = 0
+        memo = [[0 for x in range(cols)] for y in range(rows)]
```
Comment (eyurko) [inline assets/questions/bomberman.js]: Please add a more descriptive name than "memo". It's not clear what this is supposed to be.

File: assets/questions/bomberman.js [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+`class AuxiliaryCode(object):
+    @classmethod
+    def negative(cls, board):
+        return -1
+    @classmethod
+    def ignoreWalls(cls, board):
+        if len(board) == 0 or len(board[0]) == 0: return 0
+        rows = len(board)
+        cols = len(board[0])
+        maxEnemy = 0
```
Comment (eyurko) [inline assets/questions/bomberman.js]: Not clear what maxEnemy is supposed to be, either. Mind elaborating?

File: assets/questions/bomberman.js [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+  starterCode: {
+    python:
+`def bomb(board):
+    return 0
+`
+  },
+  auxiliaryCode: {
+    python:
+`class AuxiliaryCode(object):
+    @classmethod
```
Comment (eyurko) [inline assets/questions/bomberman.js]: Add a new line character.

File: assets/questions/bomberman.js [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+                    right += 1
+                    down[c] += 1
+                else:
+                    memo[r][c] += right + down[c]
+                    if memo[r][c] > maxEnemy:
+                        maxEnemy = memo[r][c]
+        return maxEnemy
+    @classmethod
+    def bombAllEnemiesOnBoard(cls, board):
+        if (len(board) == 0 or len(board[0]) == 0): return 0
```
Comment (eyurko) [inline assets/questions/bomberman.js]: Again, do not use explicit length comparison in Python.

File: assets/questions/bomberman.js (hunk 3 above)
Comment (eyurko) [inline assets/questions/bomberman.js]: are string -> are strings

File: assets/questions/bomberman.js (hunk 4 above)
Comment (eyurko) [inline assets/questions/bomberman.js]: , and "" represents an empty cell.

File: assets/questions/bomberman.js [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+                    count += 1
+        return count
+`
+  },
+  tasks: [{
+    instructions: [
+      [
+        'For this question, you will be given a borad, in which all elements are string. ',
+        'X represent a wall, e represent an enemy, "" represent an empty cell. ',
+        'You can place an bomb in an empty cell, which will kill enemies in vertical and horizontal ',
```
Comment (eyurko) [inline assets/questions/bomberman.js]: will kill all enemies

File: assets/questions/bomberman.js [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+      },
+      {
+        input: [["", "e", ""], ["e", "", "e"], ["", "e", ""]],
+        allowedOutputs: [4]
+      }
+    ],
+    buggyOutputTests: [{
+      buggyFunctionName: 'AuxiliaryCode.ignoreWalls',
+      messages: [[
+        "You cannot kill an enemy behind a wall x. ",
```
Comment (eyurko) [inline assets/questions/bomberman.js]: [
"Try running your code on {{EXAMPLE WITH WALLS}}. Is the result returned what you expect?",
"Are there situations in which you cannot kill an enemy?",
"You seem to be ignoring walls in your solution, which is incorrect. If a bomb is placed, the resulting explosion should only go until it reaches a wall."
]

File: assets/questions/bomberman.js [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+      buggyFunctionName: 'AuxiliaryCode.bombAllEnemiesOnBoard',
+      messages: [[
+        "You cannot simply kill all enemies on the board. ",
+        "Your bomb can only kill enemies in the same row and column. ",
+        "Stop your fire once it meets a wall."
+      ].join('')
+      ]
+    },
+    {
+      buggyFunctionName: 'AuxiliaryCode.negative',
```
Comment (eyurko) [inline assets/questions/bomberman.js]: I do not think this is a useful test case -- do you expect that many students will attempt to submit this as a valid solution?

File: assets/questions/bomberman.js [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+      messages: [[
+        "You cannot kill an enemy behind a wall x. ",
+        "Stop your fire once it meets a wall."
+      ].join('')
+      ]
+    },
+    {
+      buggyFunctionName: 'AuxiliaryCode.bombAllEnemiesOnBoard',
+      messages: [[
+        "You cannot simply kill all enemies on the board. ",
```
Comment (eyurko) [inline assets/questions/bomberman.js]: [
"Try running your code on {{EXAMPLE WITH MANY ENEMIES}}. Is the result returned what you expect?",
"Are there limitations to how your bombs will kill enemies on the board?",
"You seem to returning all enemies on the board which is generally incorrect. If a bomb is placed, the resulting explosion should only go within the same row and column until it reaches a wall."
]

File: assets/questions/bomberman.js~ [hunk 17]
```diff
@@ -0,0 +1,147 @@
+// Copyright 2017 The TIE Authors. All Rights Reserved.
```
Comment (eyurko) [inline assets/questions/bomberman.js~]: This file should be removed from the commit.

File: assets/questions/bomberman.js [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+      ]
+    },
+    {
+      buggyFunctionName: 'AuxiliaryCode.negative',
+      messages: [[
+        "Don't just return -1, do something"
+      ].join('')
+      ]
+    }],
+    performanceTests: []
```
Comment (eyurko) [inline assets/questions/bomberman.js]: Please either add a performance test or a comment explaining why there are no performance tests for this question. Note that it's fine to not have performance tests, currently, as we only support linear runtimes at the moment.

File: assets/question_sets/strings.js [hunk 19]
```diff
@@ -26,5 +26,5 @@ globalData.questionSets['strings'] = {  // eslint-disable-line dot-notation
       "I will provide you with feedback."
     ].join('')
   ],
-  questionIds: ['reverseWords', 'parens', 'i18n', 'rle']
+  questionIds: ['reverseWords', 'parens', 'i18n', 'rle', 'bomberman']
```
Comment (eyurko) [inline assets/question_sets/strings.js]: No need to add this to strings, right now -- we won't be using it for our user study, and that's a curated set of problems.

File: assets/questions/bomberman.js (hunk 3 above)
Comment (jianhaolyu) [inline assets/questions/bomberman.js]: Thanks!

File: assets/question_sets/strings.js (hunk 19 above)
Comment (jianhaolyu) [inline assets/question_sets/strings.js]: Thanks! The problems have been fixed, I'm now working on CLA issue. This branch may be deleted in order to pass the CLA check.

Comment (jianhaolyu) [inline assets/question_sets/strings.js]: @eyurko

**Pull request closed without merging**